test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,386 @@
+# Copyright(c) 2020 Intel Corporation. All rights reserved.
+
+set -e
+
+##
+## Case Name: check-volume-levels
+## Preconditions:
+##    topology is nocodec with loopack in PCM0P -> PCM0C
+##    aplay should work
+##    arecord should work
+##    PCM0C capture PGA supports -50 - +30 dB range
+##    PCM0C capture PGA supports mute switch
+## Description:
+##    Set volume and mute switch to various values, measure
+##    volume gain from the actual levels and compare to set gains.
+## Case step:
+##    1. Start aplay
+##    2a. Capture 1st wav file
+##    2b. Capture 2nd wav file
+##    2c. Capture 3rd wav file
+##    3. Measure volume gains
+## Expect result:
+##    command line check with $? without error
+##
+
+# shellcheck source=case-lib/lib.sh
+source "$(dirname "${BASH_SOURCE[0]}")/../case-lib/lib.sh"
+
+OPT_OPT_lst['t']='tplg'     OPT_DESC_lst['t']='tplg file, default value is env TPLG: $''TPLG'
+OPT_PARM_lst['t']=1         OPT_VALUE_lst['t']="$TPLG"
+
+func_opt_parse_option "$@"
+
+tplg=${OPT_VALUE_lst['t']}
+
+TPLGREADER=../tools/sof-tplgreader.py
+PCM_ID=0
+CAP_FORMAT="S16_LE"
+CAP_RATE="48000"
+VOL_P30DB=80
+VOL_P10DB=60
+VOL_0DB=50
+VOL_M10DB=40
+VOL_M20DB=30
+VOL_M30DB=20
+VOL_MAX=$VOL_P30DB
+VOL_MIN=1
+VOL_MUTE=0
+VOL_NOM=$VOL_0DB
+APLAY_WAV=$(mktemp --suffix=.wav)
+ARECORD_WAV1=$(mktemp --suffix=.wav)
+ARECORD_WAV2=$(mktemp --suffix=.wav)
+ARECORD_WAV3=$(mktemp --suffix=.wav)
+
+#
+# Main test procedure
+#
+
+main () {
+    # Preparations
+    do_preparations
+    generate_sine
+
+    # Start sine playback
+    aplay_helper "$APLAY_WAV" &
+    pause
+
+    # Do capture tests
+    run_test_1
+    run_test_2
+    run_test_3
+
+    # Stop sine playback and cleanup. %1 is the remaining aplay process.
+    pause
+    dlogi "The test procedure is now complete. Killing the aplay process."
+    kill %1
+
+    # Measure, delete unnecessary wav files if passed
+    if measure_levels; then
+	rm -f "$APLAY_WAV" "$ARECORD_WAV1" "$ARECORD_WAV2" "$ARECORD_WAV3"
+    fi
+}
+
+#
+# Test #1 volume and mute switch controls
+#
+
+#  0...1s channels 1-4 max gain
+#  1...2s channels 1-4 different gains
+#  2...3s channels 1-4 nominal gain
+#  3...4s channels 1-4 muted
+#  4...5s channels 1-4 nominal gain
+#  5...7s channels 1-4 muted
+#  7...8s channels 1-4 max gain
+#  8...9s channels 1-4 muted gain
+#  9..10s channels 1-4 min gain
+# 10..11s channels 1-4 different gains
+run_test_1 () {
+    cset_max
+    cset_unmute
+    pause
+    arecord_helper "$ARECORD_WAV1" "$CAP_CHANNELS" 11 &
+    pause
+
+    cset_volume_diff1 "$CAP_CHANNELS"
+    pause
+
+    cset_nom
+    pause
+
+    cset_mute
+    pause
+
+    cset_unmute
+    pause
+
+    cset_mute
+    pause
+
+    cset_max
+    pause
+
+    cset_unmute
+    pause
+
+    cset_mutevol
+    pause
+
+    cset_min
+    pause
+
+    cset_volume_diff2 "$CAP_CHANNELS"
+
+    # Wait for arecord process to complete. %1 is aplay and %2 is the just launched arecord.
+    dlogi "Waiting"
+    wait %2
```

Comment:
I tried clean and nice looking wait %% but it didn't work (hope I didn't have any other mess). It would be possible to grep the correct %N from jobs output but I didn't do that yet. Pid would be nicer.

Reply:
Good that you were persistent. It seems to work and I must have messed up with my previous try.